subprojects/docs/src/docs/userguide/developing-plugins/writing_in_kotlin.adoc

```diff
@@ -0,0 +1,329 @@
+}
+----
+
+For similar reasons, any transitive dependency of your plugin needs to use a `kotlin-stdlib` version compatible with the Gradle one. While you can ensure that at compile time for your own code, it's often more complicated for dependencies so make sure to test your plugins and/or double check your transitive dependencies.
+
+Note: The `kotlin-stdlib` version is independent of your Kotlin Gradle Plugin version. You can use a newer Kotlin Gradle Plugin version with an older `kotlin-stdlib` using the https://kotlinlang.org/docs/compiler-reference.html#api-version-version[`apiVersion`] compiler argument.
+
+TIP: Because it is provided by Gradle at runtime, `kotlin-stdlib` must be a `compileOnly` dependency.
+
+[[sec:kotlin_reflect]]
+=== kotlin-reflect
+
+[source,kotlin]
+----
+dependencies {
+    compileOnly("org.jetbrains.kotlin:kotlin-reflect:$version")
+}
+----
+
+Not all plugins need `kotlin-reflect` but if yours does, the situation is the same as for `kotlin-stdlib`. `kotlin-reflect` needs to be compatible with the https://docs.gradle.org/current/userguide/compatibility.html#kotlin[target Gradle version]
+
+TIP: Because it is provided by Gradle at runtime, `gradle-api` must be a `compileOnly` dependency.
+
+[[sec:others]]
+=== others (`implementation` or `api`)
+
+You can add other non-compileOnly dependencies like in other JVM projects:
+
+[source,kotlin]
+----
+dependencies {
+    // Kotlin dependencies
+    // ⚠️ make sure they do not depend on an incompatible version of kotlin-stdlib
+    implementation("com.squareup:kotlinpoet:1.12.0")
+
+    // Java dependencies
+    implementation("asm:asm:3.3.1")
+}
+----
+
+These dependencies are not provided by Gradle at runtime and should use the `implementation` configuration (or `api` if part of your plugin API). Because Gradle may load different plugins with conflicting dependencies in the same https://dev.to/autonomousapps/build-compile-run-a-crash-course-in-classpaths-f4g[classloader], your plugin users may experience crashes if symbols are incompatible or missing. To mitigate those, you can relocate your dependencies.
+
+[[sec:compiler_options]]
+== Different compiler configurations
+
+You can choose different options to either improve the developer experience of writing plugins and/or improve compatibility with Gradle and Java in general.
+
+[[sec:sam_with_receiver]]
+=== `sam-with-receiver` compiler plugin
+
+Gradle comes with the https://docs.gradle.org/current/javadoc/org/gradle/api/HasImplicitReceiver.html[@HasImplicitReceiver] annotation. When used in conjunction with Kotlin https://kotlinlang.org/docs/sam-with-receiver-plugin.html[sam-with-receiver compiler plugin], it can turn the parameter of https://kotlinlang.org/docs/fun-interfaces.html[Single Abstract Method (SAM)] interfaces like https://docs.gradle.org/current/javadoc/org/gradle/api/Action.html[Action] into implicit receiver making your code look more like a DSL:
```

Comment:
❓ This should be the default which would remove the need to mention this here. It brings both DSL closer to a standard "Gradle DSL". I'm not much of a Kotlin developer so reading this is a bit confusing as to how, as a user, I can always get the "sam-with-receiver" behaviour.

Reply:
The `sam-with-receiver` is a very confusing default to me and judging by the amount of questions I see on this, I'm not the only one. As a Kotlin developer, I have the expectation that if I'm writing a `.kt` file then I should be able to use my usual Kotlin syntax. Kotlin makes it very easy to change a lambda parameter to a receiver (with `with` or `apply`) so things like sam-with-receiver and/or assignement compilers plugin contributes to making the process of writing plugins feel more magical than need be.

Of course if users want to use `sam-with-receiver` it's fine but I think it's important to be aware of the alternative.

Reply:
The Gradle team should clarify if one is preferred in the future and only allow one (if that makes sense).

Reply:
Hm, how instead of proposing `sam-with-receiver` usage - propose to add two methods in plugin api:
```
// Groovy/Java compatible
fun configureSomething(conf: Action<Something>) { .. }

// Kotlin compatible
fun configureSomething(conf: Something.() -> Unit) { ... }
```